src/librustc_typeck/check/compare_method.rs

```diff
@@ -1185,16 +1185,18 @@ fn compare_projection_bounds<'tcx>(
     impl_ty_span: Span,
     impl_trait_ref: ty::TraitRef<'tcx>,
 ) -> Result<(), ErrorReported> {
-    let is_gat = !tcx.generics_of(impl_ty.def_id).params.is_empty();
-    if impl_ty.defaultness.is_final() && !is_gat {
+    let have_gats = tcx.features().generic_associated_types;
+    if impl_ty.defaultness.is_final() && !have_gats {
         // For "final", non-generic associate type implementations, we
         // don't need this as described above.
         return Ok(());
     }
 
     let param_env = tcx.param_env(impl_ty.def_id);
 
-    let impl_substs = InternalSubsts::identity_for_item(tcx, impl_ty.container.id());
+    let impl_ty_substs = InternalSubsts::identity_for_item(tcx, impl_ty.def_id);
```

Comment:
I think it'd be nice to have a comment here explaining what `impl_ty_substs` and `rebased_substs` represent in terms of an example.

Reply:
i.e.,

given

```
impl<A, B> Foo for (A, B) {
    type Bar<C> =...
}
```

* `impl_substs` would be `[A, B, C]` placeholders
* `rebased_substs` would be `[(A, B), C]`, combining the substs from the *trait* with the generic associated type parameters